rust/src/nasl/builtin/ssh/russh/session.rs

```diff
@@ -0,0 +1,235 @@
+    // both of which feel like much worse alternatives.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new(
+        id: SessionId,
+        ip_addr: IpAddr,
+        port: Port,
+        timeout: Option<Duration>,
+        keytype: Vec<key::Name>,
+        csciphers: Vec<cipher::Name>,
+        scciphers: Vec<cipher::Name>,
+        socket: Option<Socket>,
+    ) -> Result<Self, SshError> {
+        if socket.is_some() {
+            error!("Using custom sockets not yet implemented.");
+            return Err(SshErrorKind::Unimplemented.with(id));
+        }
+        let preferred = construct_preferred(keytype, csciphers, scciphers);
+        let config = client::Config {
+            inactivity_timeout: timeout,
+            preferred,
+            ..Default::default()
+        };
+
+        let config = Arc::new(config);
+        let sh = Client {};
+
+        let session = connect(config, (ip_addr, port), sh)
+            .await
+            .map_err(|e| SshErrorKind::Connect.with(id).with(e))?;
+
+        Ok(Self { session, id })
+    }
+
+    pub async fn exec_ssh_cmd(&self, command: &str) -> Result<(String, String), SshError> {
+        let stdout = self.call(command).await.map_err(|e| {
+            SshErrorKind::RequestExec(command.to_string())
+                .with(self.id)
+                .with(e)
+        })?;
+        // TODO implement stderr properly.
+        let stderr = String::new();
+        Ok((stdout, stderr))
+    }
+
+    pub async fn call(&self, command: &str) -> Result<String, russh::Error> {
+        let mut channel = self.session.channel_open_session().await?;
+        channel.exec(true, command).await?;
+
+        let mut code = None;
+        let mut stdout = String::new();
+
+        loop {
+            // There's an event available on the session channel
+            let Some(msg) = channel.wait().await else {
+                break;
+            };
+            match msg {
+                // Write data to the terminal
+                ChannelMsg::Data { ref data } => {
+                    stdout.push_str(&bytes_to_str(data));
+                }
+                // The command has returned an exit code
+                ChannelMsg::ExitStatus { exit_status } => {
+                    code = Some(exit_status);
+                    // cannot leave the loop immediately, there might still be more data to receive
+                    channel.eof().await?;
+                }
+                _ => {}
+            }
+        }
+        if code.is_none() {
+            warn!("Program did not exit cleanly: {}", command);
+        }
+        Ok(stdout.to_string())
+    }
+
+    pub async fn auth_password(&mut self, login: &str, password: &str) -> Result<(), SshError> {
+        self.session
+            .authenticate_password(login, password)
+            .await
+            .map_err(|_| SshErrorKind::UserAuthPassword.with(self.id))
+            .map(|_| ())
+    }
+
+    pub async fn auth_public_key(
```

Comment:
q: do we support that currently? When I look at https://docs.greenbone.net/API/OSP/osp-22.4.html#element_credential or https://github.com/greenbone/openvas-scanner/blob/main/rust/src/models/credential.rs  i don't see a possibility to provide key pairs?

Reply:
Good point. It existed in the original code, so I kept it around in the refactoring and then wanted to add it in russh as well but didn't really know how to implement it. Maybe @jjnicola knows whether this should exist or not. I'll get rid of it if it shouldnt.